docs/design/sql/09-optimizer.md

```diff
@@ -0,0 +1,768 @@
+for every logical plan. As a result, we will have to consider `N * M` plans. Instead, we perform the logical 
+optimization in a separate step, extract only one best plan from the search space, and then apply the physical rules
+to on the next stage. This way, we have to consider only `N + M` plans, that alleviates the inefficiency of the core
+search algorithm of `VolcanoPlanner`.
+
+The fundamental observation, is that splitting optimization into several phases doesn't guarantee the optimal plan 
+in the general case. That is, if we generated plans `P1, P2 ... Pm ...`, and picked `Pm` as the best one, it
+doesn't mean that applying additional rules to `Pm` will produce the optimal plan `Pm'`. Another plan `Pn`, 
+such that `cost(Pn) > cost(Pm)`, could yield better plan `Pn'`, such that `cost(Pn') < cost(Pm')`.
+
+Therefore, the logical phase should generally include rules, that produce plans that are generally the best starting
+points for the subsequent physical optimization with high probability. Operator fusion, removal of unused operators, scan 
+field trimming, and filter pushdowns produces better plans in almost all cases, this is why we execute them at this 
+stage.
+
+As a part of the logical optimization process, we convert the Calcite operators with the convention `Convention.NONE` 
+to our own logical operators with the convention `HazelcastConvention.LOGICAL`. We do this through a special conversion
+rules that extend Calcite's `ConverterRule`.  
+
+If there is a Calcite operator in the tree that doesn't have a logical counterpart, it indicates that there is either an 
+unsupported operation, that we missed during validation phase, or that we missed some conversion rule. An exception will be 
+thrown in this case.
+
+*Table 5: Logical Conversion Rules*
+
+| Name | From (Calcite) | To (Hazelcast) |
+|---|---|---|
+| `MapScanLogicalRule` | `LogicalTableScan` | `MapScanLogicalRel` |
+| `FilterLogicalRule` | `LogicalFilter` | `FilterPhysicalRel` |
+| `ProjectLogicalRule` | `LogicalProject` | `ProjectLogicalRel` |
+| `ValuesLogicalRule` | `LogicalValues` | `ValuesLogicalRel` |
+
+Last, we manually add a special `RootLogicalRel` on top of the result. This operator is an abstraction of a user query
+cursor that returns results on the initiator member. 
+
+The result of the logical optimization is an optimized tree of operators with the `HazelcastConvention.LOGICAL` convention, 
+that has `RootLogicalRel` at the root.
+
+### 3.5 Physical Optimization
+
+The goal of the physical optimization is to find the physical implementations of logical operators. Some physical operators
+have 1-to-1 mapping to their logical counterparts (e.g. project), while others may have completely different implementations 
+(e.g. index scan created out of logical table scan). 
+
+#### 3.5.1 Distribution Trait
+
+Since Hazelcast stores data on multiple nodes, we should take the data distribution in count during planning. This is achieved
```

Comment:
```suggestion
Since Hazelcast stores data on multiple nodes, we should take the data distribution into account during planning. This is achieved
```